File: 2024/7/main.clj
(require '[clojure.string :as str])

(def data (slurp (first *command-line-args*)))

(def lines (str/split data #"\n"))

(defn parse-line [line]
  (let [parts (str/split line #": ")
        target (Integer/parseInt (first parts))
        nums (map Integer/parseInt (str/split (second parts) #" "))]
    [target nums]))

(defn generate-arrangements [num-slots options]
  (if (zero? num-slots)
    '(())
    (for [arrangement (generate-arrangements (dec num-slots) options)
          option options]
      (conj arrangement option))))

(defn calc-num [cur [num op]]
  (case op
    0 (+ cur num)
    1 (* cur num)
    2 (Integer/parseInt (str cur num))))

(defn calc-val [target nums op-ords]
  (let [val (reduce calc-num
                    (first nums)
                    (map vector (rest nums) op-ords))]
    (if (= val target)
      val
      nil)))

(defn calc-ops [ops [target nums]]
  (map #(calc-val target nums %)
       (generate-arrangements (dec (count nums)) ops)))

(defn check-equations [lines, ops]
  (->> (map parse-line lines)
       (map #(calc-ops ops %))
       (map #(some identity %))
       (filter some?)
       (reduce +)))

(println (check-equations lines [0 1]))

(println (check-equations lines [0 1 2]))

File: 2024/7/main.py
import itertools
import sys

data = open(sys.argv[1]).read().strip()
lines = data.split('\n')

ans = 0
ans2 = 0

for line in lines:
    parts = line.split(": ")
    target = int(parts[0])
    nums = [int(n) for n in parts[1].split()]
    num_ops = len(nums) - 1
    op_ords = itertools.product([0, 1], repeat=num_ops)
    for op_ord in op_ords:
        val = nums[0]
        for i, op in enumerate(op_ord):
            match op:
                case 0: val += nums[i + 1]
                case 1: val *= nums[i + 1]
        if val == target:
            ans += val
            break
    op_ords2 = itertools.product([0, 1, 2], repeat=num_ops)
    for op_ord in op_ords2:
        val = nums[0]
        for i, op in enumerate(op_ord):
            match op:
                case 0: val += nums[i + 1]
                case 1: val *= nums[i + 1]
                case 2: val = int(str(val) + str(nums[i + 1]))
        if val == target:
            ans2 += val
            break
print(ans)
print(ans2)
